### PR TITLE
Cleanup asserts - replace with invariants

### PR DESCRIPTION
**Pull request by sonodtt**
Remove all `assert()` statements, replacing with appropriate `INVARIANT`, `PRECONDITION`, etc.

This file has no throw statements - no cleanup required w.r. to throw statements.

### CODE REVIEW

File: src/util/ieee_float.cpp [hunk 1]
```diff
@@ -47,8 +48,11 @@ void ieee_float_spect::from_type(const floatbv_typet &type)
 {
   std::size_t width=type.get_width();
   f=type.get_f();
-  assert(f!=0);
-  assert(f<width);
+  DATA_INVARIANT(f != 0, "Mantissa must be at least 1 bit");
```
Comment (tautschnig) [inline src/util/ieee_float.cpp]: Please start error messages with a lowercase letter. (Applies across this PR.)

File: src/util/ieee_float.cpp [hunk 2]
```diff
@@ -549,15 +553,23 @@ void ieee_floatt::align()
   {
     exponent_offset-=(spec.f-lowPower2);
 
-    assert(fraction*power(2, (spec.f-lowPower2))>=f_power);
-    assert(fraction*power(2, (spec.f-lowPower2))<f_power_next);
+    INVARIANT(
+      fraction * power(2, (spec.f - lowPower2)) >= f_power,
+      "Normalisation result >= lower bound");
```
Comment (tautschnig) [inline src/util/ieee_float.cpp]: Missing verb: "should it be" or "was it found to be"?

File: src/util/ieee_float.cpp [hunk 3]
```diff
@@ -859,7 +870,10 @@ ieee_floatt &ieee_floatt::operator+=(const ieee_floatt &other)
     _other.exponent=exponent;
   }
 
-  assert(exponent==_other.exponent);
+  INVARIANT(
+    exponent == _other.exponent,
+    "Prior block equalises the exponents by setting both to the "
+    "minimum of their previous values while adjusting the mantissa");
```
Comment (tautschnig) [inline src/util/ieee_float.cpp]: I'd remove this, it seems trivially true?

**Review by tautschnig — APPROVED** [review]
A couple of nit picks, looks really good and much cleaner to me!

File: src/util/ieee_float.cpp (hunk 3 above)
Comment (hannes-steffenhagen-diffblue) [reply src/util/ieee_float.cpp]: @tautschnig Ah yes, but this being true is important for the following code, that's why its being documented here. Of course its easy to verify that the code does currently establish that condition (which is a good thing), but that doesn't mean we shouldn't document that its meant to do that. The message is a bit clumsy though (my fault, I couldn't think of anything better at the time).

**Review by hannes-steffenhagen-diffblue — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: d6ed9ec1d852a590f62e007379e89b25a33a30d4).
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

File: src/util/ieee_float.cpp (hunk 1 above)
Comment (sonodtt) [reply src/util/ieee_float.cpp]: Done.

File: src/util/ieee_float.cpp (hunk 2 above)
Comment (sonodtt) [reply src/util/ieee_float.cpp]: Done, e.g. "normalisation result must be >= lower bound".

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: ea2c66d18b0535f1f3fea1576384aeca05c9af24).
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 93699c0476e7e35b038041cda03e1b3e83266f60).
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: a52631f9c12bdbc05c383f610a1e310687f12cb6).
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: a52631f9c12bdbc05c383f610a1e310687f12cb6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/83831009

**Pull request merged**